design/service-binding.md

```diff
@@ -0,0 +1,356 @@
+
+What would be better is to annotate the Strimzi objects in a way that enabled the Service Binding Operator to populate the environment variables itself.
+
+## Service Binding Specification
+
+As an enhancement to the Service Binding Operator, the [Service Binding Specification](https://github.com/application-stacks/service-binding-specification) is being developed by the community to describe how services can be made bindable, which essentially means providing binding data and a description of where to find it.
+
+Using the service binding specification annotations with Strimzi is still a bit fiddly. Here's an example of a `Kafka` CR annotated to enable service binding.
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    servicebinding/secret/host: |-
+      {{- range .status.listeners -}}
+        {{- if and (eq .type "plain") (gt (len .addresses) 0) -}}
+          {{- with (index .addresses 0) -}}
+            {{ .host }}
+          {{- end -}}
+        {{- end -}}
+    {{- end -}}
+    servicebinding/secret/port: |-
+      {{- range .status.listeners -}}
+        {{- if and (eq .type "plain") (gt (len .addresses) 0) -}}
+          {{- with (index .addresses 0) -}}
+            {{ .port }}
+          {{- end -}}
+        {{- end -}}
+    {{- end -}}
+spec:
+  kafka:
+    version: 2.4.0
+    replicas: 1
+    listeners:
+      plain: {}
+    storage:
+      type: ephemeral
+    zookeeper:
+      replicas: 1
+      storage:
+      type: ephemeral
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost.example.com
+      port: 9092
+```
+
+The Service Binding Operator looks for annotations starting `servicebinding`.
+
+This has the same shortcomings as the first example, with the advantage that the template code is not required in every single `ServiceBindingRequest`.
+
+It would be preferable if Strimzi custom resources were appropriately annotated without the user needing to do it.
+
+## Desired way of consuming binding information
+
+The aim is to make it easy to bind to any service using a `ServiceBindingRequest` CR. Here's an example:
+
+``` yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: my-binding
+spec:
+  services:
+  - group: kafka.strimzi.io
+    kind: Kafka
+    version: v1beta1
+    resourceRef: my-cluster
+```
+
+Then the binding information is supplied in in a standard way, such as environment variables. Of course, in combination with developer tooling, this can remove the need for the developer to write the code to read credentials and so on. Instead, they're presented to the application as environment variables with well known names.
+
+## Binding data
+
+To connect to Strimzi, a service binding needs the followng:
+
+* **host** - hostname or IP address
+* **port** - port number
+* **userName** - username, optional
+* **password** - password or token, optional
+* **certificate** - certificate, optional
```

Comment:
it seems that the service binding specification doesn't consider the use case of TLS client authentication which a user can enable in Strimzi. The certificate field should be just the one used for TLS encryption (the cluster CA certificate) while allowing authentication just with username and password. I guess it's going to be a well-known limitation in this case.

Reply:
Maybe using the `KafkaUser` you mentioned later is the way to overcome this problem, right?

Reply:
I think you're right about the limitation. I'd thought about it to. I can put another PR into the Service Binding Spec to get it enhanced such that it can handle all of the Strimzi options for certificates and credentials. There's nothing really unusual in the way Strimzi works here, but it does offer a few options which the SBO ought to be able to accommodate.

Reply:
Document now updated heavily including support for TLS.